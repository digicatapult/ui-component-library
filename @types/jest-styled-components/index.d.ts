declare module 'jest-styled-components/serializer'
