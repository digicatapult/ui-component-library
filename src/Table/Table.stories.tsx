import React from 'react'
import styled from 'styled-components'

import PadlockIcon from '../Icons/Padlock.js'
import ClockIcon from '../Icons/Clock.js'
import CertificateIssueIcon from '../Icons/Certificate-Issued.js'
import CertificateRevokedIcon from '../Icons/Certificate-Revoked.js'
import CertificateCO2Icon from '../Icons/Certificate-CO2.js'
import Table, { TableProps } from './index.js'

export default {
  title: 'Components/Table',
  component: Table,
  argTypes: {
    headers: {},
    rows: {},
  },
}

const Row = styled('div')`
  display: flex;
  flex-direction: row;
  gap: 5px;
  align-items: center;
`

const Col = styled('div')`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`

const Title = styled('div')`
  color: #6aa685;
  font-family: Inter;
  font-size: 16px;
  font-style: normal;
  font-weight: 700;
  line-height: 0px; /* 0% */
`

const Status = styled('div')`
  width: 90px;
  background-color: ${(props) => props.color || '#27847a'};
  color: #fff;
  border-radius: 60px;

  text-align: center;
  font-family: Roboto;
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 10px; /* 83.333% */
`

const Date = styled('div')`
  color: #27847a;

  font-family: Inter;
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 10px; /* 83.333% */
`

const NameCell = ({ name, date }) => (
  <Col>
    <Row>
      <PadlockIcon />
      <Title>{name}</Title>
    </Row>
    <Row>
      <ClockIcon />
      <Date>{date}</Date>
    </Row>
  </Col>
)

const StatusCell = ({ color = '#27847a', Icon, status }) => (
  <Col>
    <Icon />
    <Status color={color}>{status}</Status>
  </Col>
)

const DefaultStoryTemplate = (args: TableProps) => <Table {...args} />

export const Hyproof = DefaultStoryTemplate.bind({})

Hyproof.args = {
  headers: [
    'Date',
    'H2 Batch size',
    'Electric energy use',
    'Electric energy use',
    'Status',
  ],
  rows: [
    [
      <NameCell name={'UK-HYPROOF-0001'} date={'2021/01/12, 22:50h'} />,
      '1.0 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell Icon={CertificateIssueIcon} status={'issued'} />,
    ],
    [
      <NameCell name={'UK-HYPROOF-0002'} date={'2024/02/22, 22:50h'} />,
      '0.5 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell Icon={CertificateCO2Icon} status={'C.Embodiment'} />,
    ],
    [
      <NameCell name={'UK-HYPROOF-0003'} date={'2024/02/02, 22:50h'} />,
      '2.0 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell Icon={CertificateCO2Icon} status={'C.Embodiment'} />,
    ],
    [
      <NameCell name={'UK-HYPROOF-0004'} date={'2021/01/12, 22:50h'} />,
      '1.0 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell
        Icon={CertificateRevokedIcon}
        color={'red'}
        status={'Revoked'}
      />,
    ],
    [
      <NameCell name={'UK-HYPROOF-0005'} date={'2024/02/22, 22:50h'} />,
      '0.5 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell Icon={CertificateIssueIcon} status={'issued'} />,
    ],
    [
      <NameCell name={'UK-HYPROOF-0006'} date={'2024/02/02, 22:50h'} />,
      '2.0 MWh',
      '4.2 MWh',
      '45g CO2e',
      <StatusCell
        Icon={CertificateRevokedIcon}
        color={'red'}
        status={'Revoked'}
      />,
    ],
  ],
}
