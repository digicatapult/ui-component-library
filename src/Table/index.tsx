import React from 'react'
import styled from 'styled-components'

export interface TableProps {
  styles?: {
    tr: React.CSSProperties
    th: React.CSSProperties
  }
  rows: RowValue[][]
  headers: string[]
}

type RowValue = React.ReactNode | number | string | null | undefined
const defaultStyle = `
  color: var(--Black, #000);
  font-family: Roboto Mono;
  font-size: 16px;
  font-style: normal;
  text-align: left;
  font-weight: 400;
  line-height: normal;
  text-decoration-line: underline;
`

const hyproof = `
  border-bottom: solid 1px #FFF;
  text-align: left;
  padding: 2px 5px;
  border-left: solid 1px #FFF;
  background-color: #27847A;
  color: #FFF;

font-family: Roboto;
font-size: 14px;
font-style: normal;
font-weight: 500;
line-height: 26x; /* 171.429% */
`

const Table: React.FC<React.PropsWithChildren<TableProps>> = (props) => (
  <Background>
    <Wrapper {...props}>
      <TR key={Math.random().toString()} style={props.styles?.tr}>
        {props.headers.map((header: string) => (
          <TH key={Math.random().toString()} style={props.styles?.th || {}}>
            {header}
          </TH>
        ))}
      </TR>
      {props.rows.map((row: RowValue[]) => (
        <TR key={Math.random().toString()} style={props.styles?.tr}>
          {row.map((value: RowValue) => (
            <TD key={Math.random().toString()}>{value}</TD>
          ))}
        </TR>
      ))}
    </Wrapper>
  </Background>
)

const TH = styled('th')`
  ${hyproof}
`
const TD = styled('td')`
  color: #1a1a1a;

  font-family: Inter;
  font-size: 26px;
  font-style: normal;
  font-weight: 500;
  line-height: 0px; /* 0% */
  padding: 5px;
  height: 60px;
  text-align: center;
`
const TR = styled('tr')`
  background-color: #fff;
  outline: thin solid #27847a;
`

const Wrapper = styled('table')<TableProps>`
  width: ${({ width }) => width || '100%'};
  border-collapse: separate;
  border-spacing: 0 10px;
`

const Background = styled('div')`
  background-color: #27847a;
  padding: 20px;
  height: 1200px;
`

export default Table
