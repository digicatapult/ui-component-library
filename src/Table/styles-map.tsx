export const styles = {
  default: {
    root: {
      borderCollapse: 'collapse',
      outline: 'thin solid #000',
    },
    td: {
      color: 'var(--Black, #000)',
      fontFamily: 'Roboto Mono',
      fontSize: '16px',
      fontStyle: 'normal',
      textAlign: 'left',
      fontWeight: 400,
      lineHeight: 'normal',
    },
    tr: {},
    th: {
      color: 'var(--Black, #000)',
      fontFamily: 'Roboto Mono',
      fontSize: '16px',
      fontStyle: 'normal',
      textAlign: 'left',
      fontWeight: 400,
      lineHeight: 'normal',
      textDecorationLine: 'underline',
    },
  },
  hyproof: {
    root: {
      borderCollapse: 'separate',
      borderSpacing: '0 10px',
    },
    tr: {
      backgroundColor: '#fff',
      outline: 'thin solid #27847a',
    },
    td: {
      color: '#1a1a1a',
      fontFamily: 'Inter',
      fontSize: '26px',
      fontStyle: 'normal',
      fontWeight: '500',
      lineHeight: '0px',
      padding: '5px',
      height: '60px',
      textAlign: 'center',
    },
    th: {
      borderBottom: 'solid 1px #fff',
      textAlign: 'left',
      padding: '2px 5px',
      borderLeft: 'solid 1px #fff',
      backgroundColor: '#27847a',
      color: '#fff',
      fontFamily: 'Roboto',
      fontSize: '14px',
      fontStyle: 'normal',
      fontWeight: '500',
      lineHeight: '26px',
    },
  },
}
