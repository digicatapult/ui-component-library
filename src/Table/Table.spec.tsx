import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Table from './index.js'

describe('Table', () => {
  test('default', () => {
    const instance = renderer.create(
      <Table
        headers={['Timestamp', 'Subject', 'HTML']}
        rows={[
          [new Date().toISOString(), 'test-1', <h1>html element</h1>],
          [new Date().toISOString(), 'test-2', <p>paragraph</p>],
          [new Date().toISOString(), 'test-3', <button>button</button>],
        ]}
      />,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
