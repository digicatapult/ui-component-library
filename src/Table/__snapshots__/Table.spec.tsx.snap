// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table default 1`] = `
.c3 {
  border-bottom: solid 1px #FFF;
  text-align: left;
  padding: 2px 5px;
  border-left: solid 1px #FFF;
  background-color: #27847A;
  color: #FFF;
  font-family: Roboto;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: 26x;
}

.c4 {
  color: #1a1a1a;
  font-family: Inter;
  font-size: 26px;
  font-style: normal;
  font-weight: 500;
  line-height: 0px;
  padding: 5px;
  height: 60px;
  text-align: center;
}

.c2 {
  background-color: #fff;
  outline: thin solid #27847a;
}

.c1 {
  width: 100%;
  border-collapse: separate;
  border-spacing: 0 10px;
}

.c0 {
  background-color: #27847a;
  padding: 20px;
  height: 1200px;
}

<div
  className="c0"
>
  <table
    className="c1"
    headers={
      [
        "Timestamp",
        "Subject",
        "HTML",
      ]
    }
    rows={
      [
        [
          "2010-01-01T00:00:00.000Z",
          "example-1",
          <p>
            paragraph
          </p>,
        ],
        [
          "2011-01-01T00:00:00.000Z",
          "example-2",
          <button>
            button
          </button>,
        ],
        [
          "2012-01-01T00:00:00.000Z",
          "example-3",
          <h3>
            heading #3
          </h3>,
        ],
      ]
    }
  >
    <tr
      className="c2"
    >
      <th
        className="c3"
        style={{}}
      >
        Timestamp
      </th>
      <th
        className="c3"
        style={{}}
      >
        Subject
      </th>
      <th
        className="c3"
        style={{}}
      >
        HTML
      </th>
    </tr>
    <tr
      className="c2"
    >
      <td
        className="c4"
      >
        2010-01-01T00:00:00.000Z
      </td>
      <td
        className="c4"
      >
        example-1
      </td>
      <td
        className="c4"
      >
        <p>
          paragraph
        </p>
      </td>
    </tr>
    <tr
      className="c2"
    >
      <td
        className="c4"
      >
        2011-01-01T00:00:00.000Z
      </td>
      <td
        className="c4"
      >
        example-2
      </td>
      <td
        className="c4"
      >
        <button>
          button
        </button>
      </td>
    </tr>
    <tr
      className="c2"
    >
      <td
        className="c4"
      >
        2012-01-01T00:00:00.000Z
      </td>
      <td
        className="c4"
      >
        example-3
      </td>
      <td
        className="c4"
      >
        <h3>
          heading #3
        </h3>
      </td>
    </tr>
  </table>
</div>
`;
